File: site/src/search/porter.ts
/**
 * The Porter stemmer (Porter 1980), ported line for line from
 * `pipeline/porter.py`, which is itself a port of Martin Porter's reference
 * implementation at tartarus.org.
 *
 * The port keeps the reference's documented departures from the paper (marked
 * DEPARTURE below), because Porter's published test vectors
 * (`tests/golden/porter/voc.txt` and `output.txt`, 23,531 words) encode the
 * behaviour of that implementation and both languages are tested against them.
 * If these two implementations ever disagree, the browser ranks differently
 * from every number printed on the page (X2).
 *
 * The algorithm assumes a lowercase ASCII word; the tokenizer guarantees that.
 * Words of two letters or fewer come back unchanged.
 */

const VOWELS = "aeiou";

/**
 * Porter's buffer: `b[0..k]` is the word, `j` a general-purpose offset set by
 * `ends` to the index of the character before a matched suffix. `b` may stay
 * longer than `k + 1`; everything reads through `k`.
 */
class Stemmer {
  private b: string;
  private k: number;
  private j = 0;

  constructor(word: string) {
    this.b = word;
    this.k = word.length - 1;
  }

  // --- the primitives of the paper ---------------------------------------

  /**
   * True when `b[i]` is a consonant. `y` is a consonant at the start of a word
   * and after a vowel, a vowel otherwise.
   */
  private cons(i: number): boolean {
    const ch = this.b[i];
    if (VOWELS.includes(ch)) return false;
    if (ch === "y") return i === 0 || !this.cons(i - 1);
    return true;
  }

  /** The measure of `b[0..j]`: the number of VC sequences in it. */
  private m(): number {
    let n = 0;
    let i = 0;
    const j = this.j;
    for (;;) {
      if (i > j) return n;
      if (!this.cons(i)) break;
      i += 1;
    }
    i += 1;
    for (;;) {
      for (;;) {
        if (i > j) return n;
        if (this.cons(i)) break;
        i += 1;
      }
      i += 1;
      n += 1;
      for (;;) {
        if (i > j) return n;
        if (!this.cons(i)) break;
        i += 1;
      }
      i += 1;
    }
  }

  /** True when `b[0..j]` contains a vowel. */
  private vowelInStem(): boolean {
    for (let i = 0; i <= this.j; i += 1) if (!this.cons(i)) return true;
    return false;
  }

  /** True when `b[i]` and `b[i-1]` are the same consonant. */
  private doublec(i: number): boolean {
    if (i < 1 || this.b[i] !== this.b[i - 1]) return false;
    return this.cons(i);
  }

  /**
   * True when `b[i-2..i]` is consonant-vowel-consonant and the last consonant
   * is not w, x or y. Used to restore a final `e`.
   */
  private cvc(i: number): boolean {
    if (i < 2 || !this.cons(i) || this.cons(i - 1) || !this.cons(i - 2)) {
      return false;
    }
    return !"wxy".includes(this.b[i]);
  }

  // --- suffix helpers -----------------------------------------------------

  /** True when `b[0..k]` ends with `s`; sets `j` to the index before it. */
  private ends(s: string): boolean {
    const length = s.length;
    if (length > this.k + 1) return false;
    if (this.b.slice(this.k - length + 1, this.k + 1) !== s) return false;
    this.j = this.k - length;
    return true;
  }

  /** Replace everything after `j` with `s`. */
  private setto(s: string): void {
    this.b = this.b.slice(0, this.j + 1) + s;
    this.k = this.j + s.length;
  }

  /** `setto(s)` when the stem still has a measure greater than zero. */
  private r(s: string): void {
    if (this.m() > 0) this.setto(s);
  }

  // --- the steps ----------------------------------------------------------

  /**
   * Plurals and past participles: caresses -> caress, ponies -> poni,
   * agreed -> agree, plastered -> plaster, hopping -> hop.
   */
  private step1ab(): void {
    if (this.b[this.k] === "s") {
      if (this.ends("sses")) this.k -= 2;
      else if (this.ends("ies")) this.setto("i");
      else if (this.b[this.k - 1] !== "s") this.k -= 1;
    }
    if (this.ends("eed")) {
      if (this.m() > 0) this.k -= 1;
    } else if ((this.ends("ed") || this.ends("ing")) && this.vowelInStem()) {
      this.k = this.j;
      if (this.ends("at")) this.setto("ate");
      else if (this.ends("bl")) this.setto("ble");
      else if (this.ends("iz")) this.setto("ize");
      else if (this.doublec(this.k)) {
        this.k -= 1;
        if ("lsz".includes(this.b[this.k])) this.k += 1;
      } else if (this.m() === 1 && this.cvc(this.k)) this.setto("e");
    }
  }

  /**
   * Terminal y -> i when the stem holds a vowel: happy -> happi and
   * enjoy -> enjoi, while sky stays sky. (The reference's later revision tests
   * the preceding letter instead; the published vectors were made with this
   * rule, so this rule is what the port keeps.)
   */
  private step1c(): void {
    if (this.ends("y") && this.vowelInStem()) {
      this.b = `${this.b.slice(0, this.k)}i${this.b.slice(this.k + 1)}`;
    }
  }

  /**
   * Double suffixes to single ones: relational -> relate (then relat in
   * step 4), hesitanci -> hesitance, digitizer -> digitize.
   */
  private step2(): void {
    if (this.k < 1) return;
    const ch = this.b[this.k - 1];
    if (ch === "a") {
      if (this.ends("ational")) this.r("ate");
      else if (this.ends("tional")) this.r("tion");
    } else if (ch === "c") {
      if (this.ends("enci")) this.r("ence");
      else if (this.ends("anci")) this.r("ance");
    } else if (ch === "e") {
      if (this.ends("izer")) this.r("ize");
    } else if (ch === "l") {
      // DEPARTURE: the paper has abli -> able; the reference uses the more
      // general bli -> ble.
      if (this.ends("bli")) this.r("ble");
      else if (this.ends("alli")) this.r("al");
      else if (this.ends("entli")) this.r("ent");
      else if (this.ends("eli")) this.r("e");
      else if (this.ends("ousli")) this.r("ous");
    } else if (ch === "o") {
      if (this.ends("ization")) this.r("ize");
      else if (this.ends("ation")) this.r("ate");
      else if (this.ends("ator")) this.r("ate");
    } else if (ch === "s") {
      if (this.ends("alism")) this.r("al");
      else if (this.ends("iveness")) this.r("ive");
      else if (this.ends("fulness")) this.r("ful");
      else if (this.ends("ousness")) this.r("ous");
    } else if (ch === "t") {
      if (this.ends("aliti")) this.r("al");
      else if (this.ends("iviti")) this.r("ive");
      else if (this.ends("biliti")) this.r("ble");
    } else if (ch === "g") {
      // DEPARTURE: not in the paper; it stems archaeologi, biologi, ...
      if (this.ends("logi")) this.r("log");
    }
  }

  /**
   * -ic-, -full, -ness and friends: triplicate -> triplic,
   * hopefulness -> hope, goodness -> good.
   */
  private step3(): void {
    const ch = this.b[this.k];
    if (ch === "e") {
      if (this.ends("icate")) this.r("ic");
      else if (this.ends("ative")) this.r("");
      else if (this.ends("alize")) this.r("al");
    } else if (ch === "i") {
      if (this.ends("iciti")) this.r("ic");
    } else if (ch === "l") {
      if (this.ends("ical")) this.r("ic");
      else if (this.ends("ful")) this.r("");
    } else if (ch === "s") {
      if (this.ends("ness")) this.r("");
    }
  }

  /**
   * Strip -ant, -ence, -ion and the rest when the stem measures above one:
   * revival -> reviv, allowance -> allow, adoption -> adopt.
   */
  private step4(): void {
    if (this.k < 1) return;
    const ch = this.b[this.k - 1];
    if (ch === "a") {
      if (!this.ends("al")) return;
    } else if (ch === "c") {
      if (!(this.ends("ance") || this.ends("ence"))) return;
    } else if (ch === "e") {
      if (!this.ends("er")) return;
    } else if (ch === "i") {
      if (!this.ends("ic")) return;
    } else if (ch === "l") {
      if (!(this.ends("able") || this.ends("ible"))) return;
    } else if (ch === "n") {
      if (
        !(
          this.ends("ant") ||
          this.ends("ement") ||
          this.ends("ment") ||
          this.ends("ent")
        )
      ) {
        return;
      }
    } else if (ch === "o") {
      // DEPARTURE: -ion goes only when preceded by s or t.
      if (this.ends("ion") && this.j >= 0 && "st".includes(this.b[this.j])) {
        // fall through to the measure test
      } else if (this.ends("ou")) {
        // fall through to the measure test
      } else return;
    } else if (ch === "s") {
      if (!this.ends("ism")) return;
    } else if (ch === "t") {
      if (!(this.ends("ate") || this.ends("iti"))) return;
    } else if (ch === "u") {
      if (!this.ends("ous")) return;
    } else if (ch === "v") {
      if (!this.ends("ive")) return;
    } else if (ch === "z") {
      if (!this.ends("ize")) return;
    } else return;
    if (this.m() > 1) this.k = this.j;
  }

  /**
   * Tidy up: a final e goes unless the stem is short, and a double l
   * collapses. probate -> probat, controll -> control, roll -> roll.
   */
  private step5(): void {
    this.j = this.k;
    if (this.b[this.k] === "e") {
      const a = this.m();
      if (a > 1 || (a === 1 && !this.cvc(this.k - 1))) this.k -= 1;
    }
    if (this.b[this.k] === "l" && this.doublec(this.k) && this.m() > 1) {
      this.k -= 1;
    }
  }

  run(): string {
    this.step1ab();
    this.step1c();
    this.step2();
    this.step3();
    this.step4();
    this.step5();
    return this.b.slice(0, this.k + 1);
  }
}

/**
 * Memoised, as the Python side is: a query repeats words the reader has
 * already typed, and the parity test stems 23,531 words twice over.
 */
const cache = new Map<string, string>();
const CACHE_LIMIT = 1 << 17;

/**
 * The Porter stem of a lowercase ASCII word. Words of two letters or fewer
 * are returned unchanged, as in the reference.
 */
export function stem(word: string): string {
  if (word.length <= 2) return word;
  const hit = cache.get(word);
  if (hit !== undefined) return hit;
  const value = new Stemmer(word).run();
  if (cache.size < CACHE_LIMIT) cache.set(word, value);
  return value;
}

File: pipeline/porter.py
"""The Porter stemmer (Porter 1980), ported from Martin Porter's own
reference implementation at tartarus.org.

This is a line-for-line port of the reference `stem.c` / `porter.py`, keeping
the documented departures from the 1980 paper that the reference makes (they
are marked DEPARTURE below), because Porter's published test vectors
(`tests/golden/porter/voc.txt` and `output.txt`, 23,531 words) encode the
behaviour of that implementation, not of the paper.

The algorithm assumes a lowercase ASCII word; the tokenizer guarantees that.
Words of two letters or fewer come back unchanged.
"""

from __future__ import annotations

from functools import lru_cache

_VOWELS = frozenset("aeiou")


class _Stemmer:
    """Porter's buffer: `b[0..k]` is the word, `j` a general-purpose offset
    set by `_ends` to the index of the character before a matched suffix."""

    def __init__(self, word: str) -> None:
        self.b = word
        self.k = len(word) - 1
        self.j = 0

    # --- the primitives of the paper -------------------------------------

    def _cons(self, i: int) -> bool:
        """True when `b[i]` is a consonant. `y` is a consonant at the start of
        a word and after a vowel, a vowel otherwise."""
        ch = self.b[i]
        if ch in _VOWELS:
            return False
        if ch == "y":
            return i == 0 or not self._cons(i - 1)
        return True

    def _m(self) -> int:
        """The measure of `b[0..j]`: the number of VC sequences in it."""
        n = 0
        i = 0
        j = self.j
        while True:
            if i > j:
                return n
            if not self._cons(i):
                break
            i += 1
        i += 1
        while True:
            while True:
                if i > j:
                    return n
                if self._cons(i):
                    break
                i += 1
            i += 1
            n += 1
            while True:
                if i > j:
                    return n
                if not self._cons(i):
                    break
                i += 1
            i += 1

    def _vowel_in_stem(self) -> bool:
        """True when `b[0..j]` contains a vowel."""
        return any(not self._cons(i) for i in range(self.j + 1))

    def _doublec(self, i: int) -> bool:
        """True when `b[i]` and `b[i-1]` are the same consonant."""
        if i < 1 or self.b[i] != self.b[i - 1]:
            return False
        return self._cons(i)

    def _cvc(self, i: int) -> bool:
        """True when `b[i-2..i]` is consonant-vowel-consonant and the last
        consonant is not w, x or y. Used to restore a final `e`."""
        if i < 2 or not self._cons(i) or self._cons(i - 1) or not self._cons(i - 2):
            return False
        return self.b[i] not in "wxy"

    # --- suffix helpers ---------------------------------------------------

    def _ends(self, s: str) -> bool:
        """True when `b[0..k]` ends with `s`; sets `j` to the index before it."""
        length = len(s)
        if length > self.k + 1:
            return False
        if self.b[self.k - length + 1 : self.k + 1] != s:
            return False
        self.j = self.k - length
        return True

    def _setto(self, s: str) -> None:
        """Replace everything after `j` with `s`."""
        self.b = self.b[: self.j + 1] + s
        self.k = self.j + len(s)

    def _r(self, s: str) -> None:
        """`_setto(s)` when the stem still has a measure greater than zero."""
        if self._m() > 0:
            self._setto(s)

    # --- the steps --------------------------------------------------------

    def _step1ab(self) -> None:
        """Plurals and past participles: caresses -> caress, ponies -> poni,
        agreed -> agree, plastered -> plaster, hopping -> hop."""
        if self.b[self.k] == "s":
            if self._ends("sses"):
                self.k -= 2
            elif self._ends("ies"):
                self._setto("i")
            elif self.b[self.k - 1] != "s":
                self.k -= 1
        if self._ends("eed"):
            if self._m() > 0:
                self.k -= 1
        elif (self._ends("ed") or self._ends("ing")) and self._vowel_in_stem():
            self.k = self.j
            if self._ends("at"):
                self._setto("ate")
            elif self._ends("bl"):
                self._setto("ble")
            elif self._ends("iz"):
                self._setto("ize")
            elif self._doublec(self.k):
                self.k -= 1
                if self.b[self.k] in "lsz":
                    self.k += 1
            elif self._m() == 1 and self._cvc(self.k):
                self._setto("e")

    def _step1c(self) -> None:
        """Terminal y -> i when the stem holds a vowel: happy -> happi and
        enjoy -> enjoi, while sky stays sky. (The reference's later revision
        tests the preceding letter instead; the published vectors were made
        with this rule, so this rule is what the port keeps.)"""
        if self._ends("y") and self._vowel_in_stem():
            self.b = self.b[: self.k] + "i" + self.b[self.k + 1 :]

    def _step2(self) -> None:
        """Double suffixes to single ones: relational -> relate (then relat in
        step 4), hesitanci -> hesitance, digitizer -> digitize."""
        if self.k < 1:
            return
        ch = self.b[self.k - 1]
        if ch == "a":
            if self._ends("ational"):
                self._r("ate")
            elif self._ends("tional"):
                self._r("tion")
        elif ch == "c":
            if self._ends("enci"):
                self._r("ence")
            elif self._ends("anci"):
                self._r("ance")
        elif ch == "e":
            if self._ends("izer"):
                self._r("ize")
        elif ch == "l":
            # DEPARTURE: the paper has abli -> able; the reference uses the
            # more general bli -> ble.
            if self._ends("bli"):
                self._r("ble")
            elif self._ends("alli"):
                self._r("al")
            elif self._ends("entli"):
                self._r("ent")
            elif self._ends("eli"):
                self._r("e")
            elif self._ends("ousli"):
                self._r("ous")
        elif ch == "o":
            if self._ends("ization"):
                self._r("ize")
            elif self._ends("ation"):
                self._r("ate")
            elif self._ends("ator"):
                self._r("ate")
        elif ch == "s":
            if self._ends("alism"):
                self._r("al")
            elif self._ends("iveness"):
                self._r("ive")
            elif self._ends("fulness"):
                self._r("ful")
            elif self._ends("ousness"):
                self._r("ous")
        elif ch == "t":
            if self._ends("aliti"):
                self._r("al")
            elif self._ends("iviti"):
                self._r("ive")
            elif self._ends("biliti"):
                self._r("ble")
        elif ch == "g":
            # DEPARTURE: not in the paper; it stems archaeologi, biologi, ...
            if self._ends("logi"):
                self._r("log")

    def _step3(self) -> None:
        """-ic-, -full, -ness and friends: triplicate -> triplic,
        hopefulness -> hope, goodness -> good."""
        ch = self.b[self.k]
        if ch == "e":
            if self._ends("icate"):
                self._r("ic")
            elif self._ends("ative"):
                self._r("")
            elif self._ends("alize"):
                self._r("al")
        elif ch == "i":
            if self._ends("iciti"):
                self._r("ic")
        elif ch == "l":
            if self._ends("ical"):
                self._r("ic")
            elif self._ends("ful"):
                self._r("")
        elif ch == "s":
            if self._ends("ness"):
                self._r("")

    def _step4(self) -> None:
        """Strip -ant, -ence, -ion and the rest when the stem measures above
        one: revival -> reviv, allowance -> allow, adoption -> adopt."""
        if self.k < 1:
            return
        ch = self.b[self.k - 1]
        if ch == "a":
            if not self._ends("al"):
                return
        elif ch == "c":
            if not (self._ends("ance") or self._ends("ence")):
                return
        elif ch == "e":
            if not self._ends("er"):
                return
        elif ch == "i":
            if not self._ends("ic"):
                return
        elif ch == "l":
            if not (self._ends("able") or self._ends("ible")):
                return
        elif ch == "n":
            if not (
                self._ends("ant")
                or self._ends("ement")
                or self._ends("ment")
                or self._ends("ent")
            ):
                return
        elif ch == "o":
            # DEPARTURE: -ion goes only when preceded by s or t.
            if self._ends("ion") and self.j >= 0 and self.b[self.j] in "st":
                pass
            elif self._ends("ou"):
                pass
            else:
                return
        elif ch == "s":
            if not self._ends("ism"):
                return
        elif ch == "t":
            if not (self._ends("ate") or self._ends("iti")):
                return
        elif ch == "u":
            if not self._ends("ous"):
                return
        elif ch == "v":
            if not self._ends("ive"):
                return
        elif ch == "z":
            if not self._ends("ize"):
                return
        else:
            return
        if self._m() > 1:
            self.k = self.j

    def _step5(self) -> None:
        """Tidy up: a final e goes unless the stem is short, and a double l
        collapses. probate -> probat, controll -> control, roll -> roll."""
        self.j = self.k
        if self.b[self.k] == "e":
            a = self._m()
            if a > 1 or (a == 1 and not self._cvc(self.k - 1)):
                self.k -= 1
        if self.b[self.k] == "l" and self._doublec(self.k) and self._m() > 1:
            self.k -= 1

    def run(self) -> str:
        self._step1ab()
        self._step1c()
        self._step2()
        self._step3()
        self._step4()
        self._step5()
        return self.b[: self.k + 1]


@lru_cache(maxsize=1 << 17)
def stem(word: str) -> str:
    """The Porter stem of a lowercase ASCII word.

    Words of two letters or fewer are returned unchanged, as in the reference.
    Results are memoised: a corpus pass stems the same few thousand types
    hundreds of thousands of times.
    """
    if len(word) <= 2:
        return word
    return _Stemmer(word).run()
